src/Tools/BuildValidator/LocalReferenceResolver.cs

```diff
@@ -0,0 +1,126 @@
+            }
+
+            return new DirectoryInfo(nugetPackageDirectory);
+        }
+
+        public Task<MetadataReference> ResolveReferenceAsync(MetadataReferenceInfo referenceInfo)
+        {
+            var path = Search(referenceInfo.Mvid, referenceInfo.Name);
+            return Task.FromResult<MetadataReference>(MetadataReference.CreateFromFile(path));
+        }
+
+        public Task<ImmutableArray<MetadataReference>> ResolveReferencesAsync(IEnumerable<MetadataReferenceInfo> references)
+        {
+            var referenceArray = references.ToImmutableArray();
+            CacheNames(referenceArray);
+
+            var files = referenceArray.Select(r => _cache[r.Mvid]);
+
+            var metadataReferences = files.Select(f => MetadataReference.CreateFromFile(f)).Cast<MetadataReference>().ToImmutableArray();
+            return Task.FromResult(metadataReferences);
+        }
+
+        public string Search(Guid mvid, string name)
+        {
+            if (_cache.TryGetValue(mvid, out var referencePath))
+            {
+                return referencePath;
+            }
+
+            foreach (var directory in _indexDirectories)
+            {
+                var foundFile = directory.GetFiles(name, SearchOption.AllDirectories).FirstOrDefault();
+                if (foundFile is null)
+                {
+                    continue;
+                }
+
+                referencePath = foundFile.FullName;
+                _cache[mvid] = referencePath;
+
+                _logger.LogTrace($"Caching [{mvid}, {referencePath}]");
+                return referencePath;
+            }
+
+            throw new KeyNotFoundException($"[{mvid}, {name}]");
+        }
+
+        public void CacheNames(ImmutableArray<MetadataReferenceInfo> names)
+        {
+            var remaining = new HashSet<MetadataReferenceInfo>(names);
+            foreach (var directory in _indexDirectories)
+            {
+                foreach (var file in directory.GetFiles("*.*", SearchOption.AllDirectories))
+                {
+                    var info = remaining.FirstOrDefault(m => m.Name == file.Name);
+                    if (info.Name == file.Name)
+                    {
+                        if (!_cache.ContainsKey(info.Mvid))
+                        {
+                            _logger.LogTrace($"Caching [{info.Mvid}, {file.FullName}]");
+                            _cache[info.Mvid] = file.FullName;
+                        }
+
+                        remaining.Remove(info);
+                    }
+                }
+            }
+        }
+
+        public static DirectoryInfo GetArtifactsDirectory()
+        {
+            var assemblyLocation = typeof(LocalReferenceResolver).Assembly.Location;
+            var binDir = Directory.GetParent(assemblyLocation);
+
+            while (binDir != null && !binDir.FullName.EndsWith("artifacts\\bin"))
```

Comment:
This should be using a `StringComparer` that matches what is in `FileNameEqualityComparer`. Essentially that type should expose a `StringComparer` instance that can be used in cases like this that will keep file name comparisons consistent throughout the code base.

Reply:
Is there a way to go from a `StringComparer` to an overload for EndsWith? I see that it takes `StringComparison`, and I _could_ expose that, but there might be a cleaner way than having to have something like below in the `FileNameEqualityComparer` 

```c# 
public static readonly StringComparison StringComparison = StringComparison.Ordinal;
public static readonly StringComparer StringComparer = StringComparer.FromComparison(StringComparison);
```

Reply:
Agree. We should just expose both of those from `FileNameEqualityComparer`